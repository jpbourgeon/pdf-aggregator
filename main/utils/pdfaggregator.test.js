const fs = require('fs-extra');
const { resolve } = require('app-root-path');
const snapshotPdfFiles = require('./__testutils__/snapshotpdffiles');
const PdfAggregator = require('./pdfaggregator');

const testbed = resolve('main/utils/__testbed__').replace(/\\/g, '/');
const defaultOptions = {
  input: `${testbed}/pdfaggregator/input`,
  level: 0,
  depth: -1,
  output: `${testbed}/pdfaggregator/output`,
  filename: '%dossiersource%_%dateiso%',
  cover: true,
  logo: `${testbed}/pdfaggregator/logo/image.jpg`,
  title: '%dossiersource%',
  subtitle: 'Version : %date%%ligne%Auteur : jpbourgeon',
  changelog: true,
  bookmarks: true,
};

describe('PDF Aggregator', () => {
  describe('the crawlFolder function', () => {
    it('should return a valid snapshot', async () => {
      expect.assertions(1);
      let tree = await PdfAggregator.crawlFolder(defaultOptions.input);
      tree = tree.map(element => ({
        ...element,
        fullPath: `[MOCKED_OS_SPECIFIC_PATH]${element.fullPath.split('__testbed__')[1]}`,
        lastModified: 'MOCKED_DATE',
      }), []);
      expect(tree).toMatchSnapshot();
    });
  });

  describe('the getFoldersToAggregate function', () => {
    it('should throw an error if the tree is empty', () => {
      expect(() => {
        PdfAggregator.getFoldersToAggregate([], defaultOptions);
      }).toThrowError('There is no folder to aggregate');
    });

    it('should return the input folder\'s path for level 0', async () => {
      expect.assertions(1);
      const expectedValue = [defaultOptions.input.split('__testbed__')[1]];
      const tree = await PdfAggregator.crawlFolder(defaultOptions.input);
      const foldersToAggregate = PdfAggregator.getFoldersToAggregate(tree, defaultOptions)
        .map(element => element.split('__testbed__')[1], []);
      expect(foldersToAggregate).toEqual(expectedValue);
    });

    it('should match the snapshot of the folders at level 1', async () => {
      expect.assertions(1);
      const tree = await PdfAggregator.crawlFolder(defaultOptions.input);
      const foldersToAggregate = PdfAggregator.getFoldersToAggregate(tree, { ...defaultOptions, level: 1 })
        .map(element => `[MOCKED_OS_SPECIFIC_PATH]${element.split('__testbed__')[1]}`, []);
      expect(foldersToAggregate).toMatchSnapshot();
    });

    it(
      'should match the snapshot of the folders at the tree\'s max depth, if the level provided is greater than it',
      async () => {
        expect.assertions(1);
        const tree = await PdfAggregator.crawlFolder(defaultOptions.input);
        const foldersToAggregate = PdfAggregator.getFoldersToAggregate(tree, { ...defaultOptions, level: Infinity })
          .map(element => `[MOCKED_OS_SPECIFIC_PATH]${element.split('__testbed__')[1]}`, []);
        expect(foldersToAggregate).toMatchSnapshot();
      },
    );
  });

  describe('the getSubTree function', () => {
    it('should throw an error if the tree is empty', () => {
      expect(() => {
        PdfAggregator.getSubTree([{ fullPath: '/file/path', depth: 5 }], '/another/path/');
      }).toThrowError('There is no file to aggregate');
    });

    it('should match the snapshot of the subtree with an unlimited depth', () => {
      const files = PdfAggregator.getSubTree([
        { fullPath: '', depth: 0 },
        { fullPath: 'file1.pdf', depth: 0 },
        { fullPath: '/file/', depth: 1 },
        { fullPath: '/file/path/', depth: 2 },
        { fullPath: '/file/path/file1.pdf', depth: 3 },
        { fullPath: '/file/path/folder1', depth: 3 },
        { fullPath: '/file/path/folder1/file1.pdf', depth: 4 },
        { fullPath: '/file/another/path', depth: 3 },
        { fullPath: '/file/another/path/file1.pdf', depth: 4 },
      ], '/file/path/');
      expect(files).toMatchSnapshot();
    });

    it('should match the snapshot of the subtree with a limited depth', () => {
      const files = PdfAggregator.getSubTree([
        { fullPath: '', depth: 0 },
        { fullPath: 'file1.pdf', depth: 0 },
        { fullPath: '/file/', depth: 1 },
        { fullPath: '/file/path/', depth: 2 },
        { fullPath: '/file/path/file1.pdf', depth: 3 },
        { fullPath: '/file/path/folder1', depth: 3 },
        { fullPath: '/file/path/folder1/file1.pdf', depth: 4 },
        { fullPath: '/file/another/path', depth: 3 },
        { fullPath: '/file/another/path/file1.pdf', depth: 4 },
      ], '/file/path/', 3);
      expect(files).toMatchSnapshot();
    });
  });

  describe('the stripEmptyFolders function', () => {
    it('should match the snapshot of the cleaned subtree', () => {
      const result = PdfAggregator.stripEmptyFolders([
        { fullPath: '', type: 'directory' },
        { fullPath: 'file1.pdf', type: 'file' },
        { fullPath: '/file/', type: 'directory' },
        { fullPath: '/file/file1.pdf', type: 'file' },
        { fullPath: '/file/path/', type: 'directory' },
        { fullPath: '/file/path/folder1', type: 'directory' },
        { fullPath: '/file/path/folder1/file1.pdf', type: 'file' },
        { fullPath: '/file/path/folder2', type: 'directory' },
        { fullPath: '/file/another/path', type: 'directory' },
      ]);
      expect(result).toMatchSnapshot();
    });
    it('should work on empty folders', () => {
      const result = PdfAggregator.stripEmptyFolders([]);
      expect(result).toMatchSnapshot();
    });
  });

  describe('the deduplicatePdfPath function', () => {
    it('should work');
  });

  describe('Tests that write to the output folder', () => {
    beforeEach(async (done) => {
      // discard the output folder's content
      await fs.emptyDir(`${testbed}/pdfaggregator/output`);
      done();
    });

    afterAll(async (done) => {
      // discard the output folder's content
      await fs.emptyDir(`${testbed}/pdfaggregator/output`);
      done();
    });

    describe('the makeEmptyPdf function', () => {
      it('should make an empty pdf document to use as a template', async () => {
        expect.assertions(1);
        const folder = `${testbed}/pdfaggregator/output`;
        await PdfAggregator.makeEmptyPdf(folder);
        const result = await snapshotPdfFiles(folder);
        expect(result).toMatchSnapshot();
      });
    });

    describe('the aggregate function', () => {
      beforeEach(async (done) => {
      // discard the output folder's content
        await fs.emptyDir(`${testbed}/pdfaggregator/output`);
        done();
      });

      afterAll(async (done) => {
      // discard the output folder's content
        await fs.emptyDir(`${testbed}/pdfaggregator/output`);
        done();
      });

      it('should work', async () => {
        await PdfAggregator.aggregate({ ...defaultOptions, level: 0, depth: -1 }, jest.fn());
      });

      describe('on an empty folder', () => {
        it('should throw an error with the following config: cover page, change log, outline');

        it('should throw an error with the following config: no cover page, no change log, no outline');
      });

      describe('on the root folder with unlimited depth', () => {
        it('should match the snapshot with the following config: cover page, change log, outline');

        it('should match the snapshot with the following config: no cover page, no change log, no outline');
      });

      describe('on level 1 folders with depth 1', () => {
        it('should match the snapshot with the following config: cover page, change log, outline');

        it('should match the snapshot with the following config: no cover page, no change log, no outline');
      });

      describe('on level 1 folders with unlimited depth', () => {
        it('should match the snapshot with the following config: cover page, change log, outline');

        it('should match the snapshot with the following config: no cover page, no change log, no outline');
      });

      it('should send localized messages (i18n)');
    });
  });
});
