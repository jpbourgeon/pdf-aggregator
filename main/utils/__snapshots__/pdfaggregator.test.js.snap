// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDF Aggregator the crawlFolder function should return a valid snapshot 1`] = `
Array [
  Object {
    "depth": 1,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/File_01.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "File_01.pdf",
    "parentDir": "",
    "type": "file",
  },
  Object {
    "depth": 1,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01",
    "parentDir": "",
    "type": "directory",
  },
  Object {
    "depth": 2,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_File_01.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_File_01.pdf",
    "parentDir": "Folder_01",
    "type": "file",
  },
  Object {
    "depth": 2,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_01",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder_01",
    "parentDir": "Folder_01",
    "type": "directory",
  },
  Object {
    "depth": 3,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_01/Folder_01_Subfolder01_File_01.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder01_File_01.pdf",
    "parentDir": "Folder_01/Folder_01_Subfolder_01",
    "type": "file",
  },
  Object {
    "depth": 3,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_01/Folder_01_Subfolder01_File_02.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder01_File_02.pdf",
    "parentDir": "Folder_01/Folder_01_Subfolder_01",
    "type": "file",
  },
  Object {
    "depth": 3,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_01/Folder_01_Subfolder_01_Subsubfolder01",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder_01_Subsubfolder01",
    "parentDir": "Folder_01/Folder_01_Subfolder_01",
    "type": "directory",
  },
  Object {
    "depth": 2,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_02",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder_02",
    "parentDir": "Folder_01",
    "type": "directory",
  },
  Object {
    "depth": 3,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_02/Folder_01_Subfolder02_File_01.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_01_Subfolder02_File_01.pdf",
    "parentDir": "Folder_01/Folder_01_Subfolder_02",
    "type": "file",
  },
  Object {
    "depth": 1,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_02",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_02",
    "parentDir": "",
    "type": "directory",
  },
  Object {
    "depth": 2,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_02/Folder_02_Subfolder_01",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_02_Subfolder_01",
    "parentDir": "Folder_02",
    "type": "directory",
  },
  Object {
    "depth": 3,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_02/Folder_02_Subfolder_01/Folder_02_Subfolder_01_File_01.pdf",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_02_Subfolder_01_File_01.pdf",
    "parentDir": "Folder_02/Folder_02_Subfolder_01",
    "type": "file",
  },
  Object {
    "depth": 2,
    "fullPath": "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_02/Folder_02_Subfolder_02",
    "lastModified": "MOCKED_DATE",
    "name": "Folder_02_Subfolder_02",
    "parentDir": "Folder_02",
    "type": "directory",
  },
]
`;

exports[`PDF Aggregator the getFoldersToAggregate function should match the snapshot of the folders at level 1 1`] = `
Array [
  "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01",
  "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_02",
]
`;

exports[`PDF Aggregator the getFoldersToAggregate function should match the snapshot of the folders at the tree's max depth, if the level provided is greater than it 1`] = `
Array [
  "[MOCKED_OS_SPECIFIC_PATH]/pdfaggregator/input/Folder_01/Folder_01_Subfolder_01/Folder_01_Subfolder_01_Subsubfolder01",
]
`;

exports[`PDF Aggregator the getSubTree function should match the snapshot of the subtree with a limited depth 1`] = `
Array [
  Object {
    "depth": 2,
    "fullPath": "/file/path/",
  },
  Object {
    "depth": 3,
    "fullPath": "/file/path/file1.pdf",
  },
  Object {
    "depth": 3,
    "fullPath": "/file/path/folder1",
  },
]
`;

exports[`PDF Aggregator the getSubTree function should match the snapshot of the subtree with an unlimited depth 1`] = `
Array [
  Object {
    "depth": 2,
    "fullPath": "/file/path/",
  },
  Object {
    "depth": 3,
    "fullPath": "/file/path/file1.pdf",
  },
  Object {
    "depth": 3,
    "fullPath": "/file/path/folder1",
  },
  Object {
    "depth": 4,
    "fullPath": "/file/path/folder1/file1.pdf",
  },
]
`;

exports[`PDF Aggregator the stripEmptyFolders function should match the snapshot of the cleaned subtree 1`] = `
Array [
  Object {
    "fullPath": "",
    "type": "directory",
  },
  Object {
    "fullPath": "file1.pdf",
    "type": "file",
  },
  Object {
    "fullPath": "/file/",
    "type": "directory",
  },
  Object {
    "fullPath": "/file/file1.pdf",
    "type": "file",
  },
  Object {
    "fullPath": "/file/path/",
    "type": "directory",
  },
  Object {
    "fullPath": "/file/path/folder1",
    "type": "directory",
  },
  Object {
    "fullPath": "/file/path/folder1/file1.pdf",
    "type": "file",
  },
]
`;

exports[`PDF Aggregator the stripEmptyFolders function should work on empty folders 1`] = `Array []`;
